ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/networkProcessor/kinemtaticsStreamingToolboxModule/input/KSTInputFBControllerStateEstimator.java

```diff
@@ -168,8 +184,273 @@ public FrameVector3DReadOnly getEstimatedCoMVelocity()
+
+         //  ORIENTATION
+
+         estimatedPose.inverseTransform(estimatedVelocity.getAngularPart(), angularVelocityBodyFrame);
+
+         orientationUpdate.setAndScale(updateDT, angularVelocityBodyFrame);
+         orientationUpdate.scaleAdd((updateDT * updateDT) / 2, angularAccelerationBodyFrame, orientationUpdate);
+
+         deltaOrientation.setRotationVector(orientationUpdate);
+
+         estimatedPose.getOrientation().append(deltaOrientation);
+
+         //  ANGULAR VELOCITY
+
+         angularVelocityBodyFrame.scaleAdd(updateDT, angularAccelerationBodyFrame, angularVelocityBodyFrame);
+         estimatedPose.transform(angularVelocityBodyFrame, estimatedVelocity.getAngularPart());
+      }
+
+      private void estimateAcceleration(FramePose3D inputPose, YoFixedFrameSpatialVector inputVelocity)
+      {
+         correctiveVelocity.set(inputVelocity);
+         correctiveVelocity.scale(velocityGain.getDoubleValue());
+
+         //LINEAR FEEDBACK
+         positionVectorError.sub(inputPose.getPosition(), estimatedPose.getPosition());
+
+         linearVelocityVectorError.sub(correctiveVelocity.getLinearPart(), estimatedVelocity.getLinearPart());
+         positionVectorError.scale(2 / (correctionDuration.getValue() * correctionDuration.getValue()));
+         linearVelocityVectorError.scale(2 / correctionDuration.getValue());
+         linearFeedbackWorldFrame.add(positionVectorError, linearVelocityVectorError);
+
+         //ANGULAR FEEDBACK
+         quaternionError.difference(estimatedPose.getOrientation(), inputPose.getOrientation());
+         quaternionError.getRotationVector(rotationVectorErrorBodyFrame); // expressed in body-frame estimated
+         angularVelocityErrorBodyFrame.sub(correctiveVelocity.getAngularPart(), estimatedVelocity.getAngularPart()); // Velocities are expressed in world frame
+         estimatedPose.inverseTransform(angularVelocityErrorBodyFrame); // This is now in the body-fixed frame
+
+         rotationVectorErrorBodyFrame.scale(2 / (correctionDuration.getValue() * correctionDuration.getValue()));
+         angularVelocityErrorBodyFrame.scale(2 / correctionDuration.getValue());
+
+         angularFeedbackBodyFrame.add(rotationVectorErrorBodyFrame, angularVelocityErrorBodyFrame);
+
+         // This is in body-frame as described by the estimatedPose
+
+         angularAccelerationBodyFrame.set(angularFeedbackBodyFrame);
+
+         YoFrameVector3D estimatedLinearAccelerationWorldFrame = estimatedAcceleration.getLinearPart();
+         YoFrameVector3D estimatedAngularAccelerationWorldFrame = estimatedAcceleration.getAngularPart();
+
+         estimatedLinearAccelerationWorldFrame.set(linearFeedbackWorldFrame);
+         estimatedPose.transform(angularAccelerationBodyFrame, estimatedAngularAccelerationWorldFrame);
+      }
+
+      @Override
+      public void predict(double time)
+      {
+         estimationsPoseVelocity(rawInputPose, rawInputVelocity);
```

Comment:
Ok now we need to figure out and to do the decay stuff. I can think of 2 solutions:
- change the calculation of the acceleration during the update method such that the velocity equals 0 at the end of the decay period. Not sure how much that'll affect the whole estimation.
- or simply decay velocity before doing the double integration in the predict method.

I'll let you play with that 😄